dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -0,0 +1,163 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "dali/pipeline/data/views.h"
+#include "dali/kernels/transpose/transpose_gpu.h"
+#include "dali/core/static_switch.h"
+#include "dali/operators/reader/numpy_reader_gpu_op.h"
+
+namespace dali {
+
+void NumpyReaderGPU::Prefetch() {
+  // We actually prepare the next batch
+  DomainTimeRange tr("[DALI][NumpyReaderGPU] Prefetch #" + to_string(curr_batch_producer_),
+                      DomainTimeRange::kRed);
+  DataReader<GPUBackend, ImageFileWrapperGPU>::Prefetch();
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // get shapes
+  for (size_t data_idx = 0; data_idx < curr_batch.size(); ++data_idx) {
+    thread_pool_.AddWork([&curr_batch, data_idx](int tid) {
+        curr_batch[data_idx]->read_meta_f();
+      });
+  }
+  thread_pool_.RunAll();
+
+  // resize the current batch
+  std::vector<TensorShape<>> tmp_shapes;
+  auto ref_type = curr_batch[0]->type_info;
+  auto ref_shape = curr_batch[0]->shape;
+  for (size_t data_idx = 0; data_idx < curr_batch.size(); ++data_idx) {
+    auto &sample = curr_batch[data_idx];
+    DALI_ENFORCE(ref_type == sample->type_info, make_string("Inconsistent data! "
+                 "The data produced by the reader has inconsistent type:\n"
+                 "type of [", data_idx, "] is ", sample->type_info.id(), " whereas\n"
+                 "type of [0] is ", ref_type.id()));
+
+    DALI_ENFORCE(ref_shape.size() == sample->shape.size(), make_string("Inconsistent data! "
+        "The data produced by the reader has inconsistent dimensionality:\n"
+        "[", data_idx, "] has ", sample->shape.size(), " dimensions whereas\n"
+        "[0] has ", ref_shape.size(), " dimensions."));
+    DALI_ENFORCE(ref_shape.size() == sample->shape.size(),
+                  make_string("The tensors in the input batch do not all have the same number of "
+                  "dimensions for sample [0]: ", ref_shape.size(), " vs [", data_idx, "]: ",
+                  sample->shape.size(), "."));
+    tmp_shapes.push_back(sample->shape);
+  }
+
+  curr_tensor_list.Resize(TensorListShape<>(tmp_shapes), ref_type);
+
+  // read the data
+  for (size_t data_idx = 0; data_idx < curr_tensor_list.ntensor(); ++data_idx) {
+    thread_pool_.AddWork([&curr_batch, &curr_tensor_list, data_idx](int tid) {
+        curr_batch[data_idx]->read_sample_f(curr_tensor_list.raw_mutable_data(),
+                                            curr_tensor_list.tensor_offset(data_idx) *
+                                            curr_tensor_list.type().size(),
+                                            curr_tensor_list.nbytes());
+      });
+  }
+  thread_pool_.RunAll();
```

Comment:
From what I can see the API we use is synchronous, I probably would expect some scheduling on CUDA stream otherwise. Is that right?

Reply:
The currently available API is fully synchronous, that is why we use this trick with the thread poll in the prefetch.